mod/dagmodifier.go

```diff
@@ -627,3 +647,315 @@ func (dm *DagModifier) GetCtx() context.Context {
+		err = json.Unmarshal(b, &m)
+		if err != nil {
+			return nil, err
+		}
+
+		inputKeys := strings.Split(string(metakeys), ",")
+
+		exists := keyIntersects(m, inputKeys)
+
+		if exists {
+			modificationRequired = true
+			// Check if inputM and m have the same key set
+			if equalKeySets(m, inputKeys) {
+				// Scenario #1: clear metadata.
+				clear = true
+			} else {
+				// Scenario #2: delete a subset of the metadata map.
+				// iterate the inputKeys to delete each key from the existing map.
+				for _, k := range inputKeys {
+					delete(m, k)
+				}
+
+				b, err = json.Marshal(m)
+				if err != nil {
+					return nil, err
+				}
+				// Create a metadata sub-DAG
+				newMetaNode, err = mdm.buildNewMetaDataDag(b)
+				if err != nil {
+					return nil, err
+				}
+			}
+		}
+	}
+
+	if !modificationRequired {
+		return root, nil
+	}
+
+	dnode := children.DataNode
+	// if clear is true, return dnode as root.
+	if clear {
+		return dnode, nil
+	}
+
+	// Attach the modified metadata sub-DAG to a new root for the BTFS file DAG.
+	fileSize, err := FileSize(dnode)
+	if err != nil {
+		return nil, err
+	}
+
+	newRoot, err := mdm.GetDb().AttachMetadataDag(dnode, fileSize, newMetaNode)
+	if err != nil {
+		return nil, err
+	}
+	mdm.GetDb().Add(newRoot)
+
+	return newRoot, nil
+}
+
+func (mdm *MetaDagModifier) buildNewMetaDataDag(metaBytes []byte) (ipld.Node, error) {
+	var superMeta help.SuperMeta
+	err := json.Unmarshal(metaBytes, &superMeta)
+	if err != nil {
+		return nil, err
+	}
+
+	r := bytes.NewReader(metaBytes)
+	if superMeta.ChunkSize == 0 {
+		superMeta.ChunkSize = 256 * 1024
+	}
+	if superMeta.MaxLinks == 0 {
+		superMeta.MaxLinks = uint64(help.DefaultLinksPerBlock)
+	}
+	// TODO: trickle format input should be set somewhere..
+
+	metaSpl := chunker.NewMetaSplitter(r, superMeta.ChunkSize)
+	metaDb := help.NewMetaDagBuilderHelper(*mdm.GetDb(), metaSpl, nil)
+	var mnode ipld.Node
+	if superMeta.TrickleFormat {
+		mnode, err = trickle.BuildNewMetaDataDag(metaDb)
+	} else {
+		mnode, err = balanced.BuildNewMetaDataDag(metaDb)
+	}
+	if err != nil {
+		return nil, err
+	}
+	return mnode, nil
+}
+
+func (mdm *MetaDagModifier) readMetadataBytes(root ipld.Node) ([]byte, error) {
+	nd, ok := root.(*mdag.ProtoNode)
+	if !ok {
+		return nil, errors.New("Expected protobuf Merkle DAG node")
+	}
+	fsn, err := ft.FSNodeFromBytes(nd.Data())
+	if err != nil {
+		return nil, err
+	}
+	if fsn.Type() != ft.TFile {
+		return nil, errors.New("Expected file type node")
+	}
+
+	_, mnode, err := ufile.CheckAndSplitMetadata(mdm.ctx, root, mdm.dagserv, false)
+	if err != nil {
+		return nil, err
+	}
+	if mnode == nil {
+		return nil, nil
+	}
+
+	r, err := uio.NewDagReader(mdm.ctx, mnode, mdm.dagserv)
+	if err != nil {
+		return nil, err
+	}
+
+	b := make([]byte, r.Size())
+	_, err = r.CtxReadFull(mdm.ctx, b)
+	if err != nil {
+		return nil, err
+	}
+
+	return b, nil
+}
+
+func intersects(m map[string]interface{}, inputM map[string]interface{}) bool {
```

Comment:
Move the following functions to utils + some simple unit tests.

Reply:
Done